In [2]:
# Files
fhand = open('mbox.txt')
print(fhand)

<_io.TextIOWrapper name='mbox.txt' mode='r' encoding='cp1252'>


In [3]:
# new line character
stuff = 'hello\nworld'
print(stuff,len(stuff))

hello
world 11


In [4]:
# to get the number of lines in the file
fhand = open('mbox.txt')
count = 0
for line in fhand:
    count = count + 1
print('Line Count:',count)

Line Count: 132045


In [5]:
# smaller files can be read into one string using read

fhand = open('mbox.txt')
inp = fhand.read()
print(len(inp))

6687002


In [6]:
# print only first 20 characters
print(inp[:20])

From stephen.marquar


In [7]:
# each call to read exhaust the resource
print(len(fhand.read()))

0


In [8]:
# print only lines that starts with prefix From:

fhand = open('mbox.txt')
count = 0
for line in fhand:
    if line.startswith('From:'):
        print(line)
        count = count + 1
print(count)

From: stephen.marquard@uct.ac.za

From: louis@media.berkeley.edu

From: zqian@umich.edu

From: rjlowe@iupui.edu

From: zqian@umich.edu

From: rjlowe@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: gsilver@umich.edu

From: gsilver@umich.edu

From: zqian@umich.edu

From: gsilver@umich.edu

From: wagnermr@iupui.edu

From: zqian@umich.edu

From: antranig@caret.cam.ac.uk

From: gopal.ramasammycook@gmail.com

From: david.horwitz@uct.ac.za

From: david.horwitz@uct.ac.za

From: david.horwitz@uct.ac.za

From: david.horwitz@uct.ac.za

From: stephen.marquard@uct.ac.za

From: louis@media.berkeley.edu

From: louis@media.berkeley.edu

From: ray@media.berkeley.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: ray@media.berkeley.edu

From: cwen@iupui.edu

From: zqian@umich.edu

From: cwen@iupui.edu

From: zqian@umich.edu

From: zqian@umich.edu

From: zqian@umich.edu

From: mmmay@indiana.edu

From: cwen@iupui.edu

From: zqian@umich.edu

From: zqian@umich.edu

Fro

From: zach.thomas@txstate.edu

From: zach.thomas@txstate.edu

From: zach.thomas@txstate.edu

From: zach.thomas@txstate.edu

From: zach.thomas@txstate.edu

From: zach.thomas@txstate.edu

From: zach.thomas@txstate.edu

From: jimeng@umich.edu

From: jimeng@umich.edu

From: arwhyte@umich.edu

From: josrodri@iupui.edu

From: arwhyte@umich.edu

From: josrodri@iupui.edu

From: stephen.marquard@uct.ac.za

From: stephen.marquard@uct.ac.za

From: gjthomas@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: gjthomas@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: ian@caret.cam.ac.uk

From: cwen@iupui.edu

From: tnguyen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: cwen@iupui.edu

From: ian@caret.cam.a

From: dlhaines@umich.edu

From: zqian@umich.edu

From: dlhaines@umich.edu

From: zqian@umich.edu

From: zqian@umich.edu

From: john.ellis@rsmart.com

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: stuart.freeman@et.gatech.edu

From: stuart.freeman@et.gatech.edu

From: zqian@umich.edu

From: jleasia@umich.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: chmaurer@iupui.edu

From: gopal.ramasammycook@gmail.com

From: dlhaines@umich.edu

From: sgithens@caret.cam.ac.uk

From: dlhaines@umich.edu

From: aaronz@vt.edu

From: dlhaines@umich.edu

From: dlhaines@umich.edu

From: cwen@iupui.edu

From: dlhaines@umich.edu

From: zqian@umich.edu

From: dlhaines@umich.edu

From: antranig@caret.cam.ac.uk

From: ian@caret.cam.ac.uk

From: zach.thomas@txstate.edu

From: dlhaines@umich.edu

From: aaron


From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: josrodri@iupui.edu

From: josrodri@iupui.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: dlhaines@umich.edu

From: dlhaines@umich.edu

From: josrodri@iupui.edu

From: zqian@umich.edu

From: arwhyte@umich.edu

From: rjlowe@iupui.edu

From: dlhaines@umich.edu

From: arwhyte@umich.edu

From: arwhyte@umich.edu

From: arwhyte@umich.edu

From: arwhyte@umich.edu

From: dlhaines@umich.edu

From: dlhaines@umich.edu

From: arwhyte@umich.edu

From: arwhyte@umich.edu

From: arwhyte@umich.edu

From: arwhyte@umich.edu

From: arwhyte@umich.edu

From: wagnermr@iupui.edu

From: zqian@umich.edu

From: zqian@umich.edu

From: mmmay@indiana.edu

From: josrodri@iupui.edu

From: gjthomas@iupui.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: mmmay@indiana.edu

From: mmmay@indiana.


From: aaronz@vt.edu

From: aaronz@vt.edu

From: jzaremba@unicon.net

From: dlhaines@umich.edu

From: chmaurer@iupui.edu

From: zach.thomas@txstate.edu

From: dlhaines@umich.edu

1797


In [9]:
# using continue to skip uninteresting lines
fhand = open('mbox.txt')
for line in fhand:
    line = line.rstrip()
    # skip our uninterestingline
    if not line.startswith('From:'):
        continue
    # process our interesting line
    print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: zqian@umich.edu
From: cwen@iupui.edu
From: zqian@umich.edu
From: zqian@umich.edu
From: zqian@umich.edu
From: mmmay@indiana.edu
From: cwen@iupui.edu
From: zqian@umich.edu
From: zqian@umich.edu
From: zqian@umich.edu
From: cwen@iupui.ed

From: stuart.freeman@et.gatech.edu
From: cwen@iupui.edu
From: mmmay@indiana.edu
From: mmmay@indiana.edu
From: ssmail@indiana.edu
From: cwen@iupui.edu
From: mmmay@indiana.edu
From: mmmay@indiana.edu
From: cwen@iupui.edu
From: wagnermr@iupui.edu
From: rjlowe@iupui.edu
From: dlhaines@umich.edu
From: jlrenfro@ucdavis.edu
From: cwen@iupui.edu
From: zqian@umich.edu
From: arwhyte@umich.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: nuno@ufp.pt
From: wagnermr@iupui.edu
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: nuno@ufp.pt
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: cwen@iupui.edu
From: cwen@iupui.edu
From: arwhyte@umich.edu
From: arwhyte@umich.edu
From: rjlowe@iupui.edu
From: wagnermr@iupui.edu
From: gjthomas@iupui.edu
From: nuno@ufp.pt
From: wagnermr@iupui.edu
From: cwen@iupui.edu
From: ian@caret.cam.ac.uk
From: cwen@iupui.edu
From: cwen@iupui.edu
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: sg

From: aaronz@vt.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: ian@caret.cam.ac.uk
From: sgithens@caret.cam.ac.uk
From: sgithens@caret.cam.ac.uk
From: zqian@umich.edu
From: zqian@umich.edu
From: zqian@umich.edu
From: gjthomas@iupui.edu
From: gjthomas@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: cwen@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: cwen@iupui.edu
From: gjthomas@iupui.edu
From: gjthomas@iupui.edu
From: cwen@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: david.horwitz@uct.ac.za
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: gopal.ramasammycook@gmail.com
From: cwen@iupui.edu
From: cwen@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: jimeng@umich.edu
From: jimeng@umich.edu
From: jimeng@umich.edu
From: jimeng@umich

From: rjlowe@iupui.edu
From: jimeng@umich.edu
From: sgithens@caret.cam.ac.uk
From: sgithens@caret.cam.ac.uk
From: aaronz@vt.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: rjlowe@iupui.edu
From: rjlowe@iupui.edu
From: rjlowe@iupui.edu
From: ray@media.berkeley.edu
From: ajpoland@iupui.edu
From: rjlowe@iupui.edu
From: ajpoland@iupui.edu
From: rjlowe@iupui.edu
From: rjlowe@iupui.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: bkirschn@umich.edu
From: gsilver@umich.edu
From: john.ellis@rsmart.com
From: aaronz@vt.edu
From: bkirschn@umich.edu
From: gopal.ramasammycook@gmail.com
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: chmaurer@iupui.edu
From: dlhaines@umich.edu
From: dlhaines@umich.edu
From: ajpoland@iupui.edu
From: aaronz@vt.edu
From: sgithens@caret.cam.ac.uk
From: aaronz@vt.edu
From: jimeng@umich.edu
From: ostermmg@whitman.edu
From: ostermmg@whitman.edu
From: jimeng@umich.edu
From: aaronz

From: gsilver@umich.edu
From: louis@media.berkeley.edu
From: bkirschn@umich.edu
From: bkirschn@umich.edu
From: ajpoland@iupui.edu
From: rjlowe@iupui.edu
From: rjlowe@iupui.edu
From: zach.thomas@txstate.edu
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: ray@media.berkeley.edu
From: ktsao@stanford.edu
From: zqian@umich.edu
From: ian@caret.cam.ac.uk
From: chmaurer@iupui.edu
From: ian@caret.cam.ac.uk
From: gjthomas@iupui.edu
From: gjthomas@iupui.edu
From: aaronz@vt.edu
From: ian@caret.cam.ac.uk
From: aaronz@vt.edu
From: aaronz@vt.edu
From: zqian@umich.edu
From: aaronz@vt.edu
From: aaronz@vt.edu
From: mbreuker@loi.nl
From: ian@caret.cam.ac.uk
From: zqian@umich.edu
From: zqian@umich.edu
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: dlhaines@umich.edu
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: ian@caret.cam.ac.uk
From: zqian@umich.edu
From: ian@caret.cam.ac.uk
From: aaronz@vt.edu
From: zqian@umich.edu
Fr

In [10]:
# find a string in the text file

fhand = open('mbox.txt')
for line in fhand:
    line = line.strip()
    # if not found
    if line.find('@uct.ac.za') == -1:
        continue
    # if found
    print(line)


From stephen.marquard@uct.ac.za Sat Jan  5 09:14:16 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to stephen.marquard@uct.ac.za using -f
From: stephen.marquard@uct.ac.za
Author: stephen.marquard@uct.ac.za
From david.horwitz@uct.ac.za Fri Jan  4 07:02:32 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
r39753 | david.horwitz@uct.ac.za | 2008-01-04 13:05:51 +0200 (Fri, 04 Jan 2008) | 1 line
From david.horwitz@uct.ac.za Fri Jan  4 06:08:27 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
From david.horwitz@uct.ac.za Fri Jan  4 04:49:08 2008
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to david.horwitz@uct.ac.za using -f
From: david.horwitz@uct.ac.za
Author: david.horwitz@uct.ac.za
From david.horw

Author: stephen.marquard@uct.ac.za
From stephen.marquard@uct.ac.za Tue Dec 18 06:33:34 2007
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to stephen.marquard@uct.ac.za using -f
From: stephen.marquard@uct.ac.za
Author: stephen.marquard@uct.ac.za
From stephen.marquard@uct.ac.za Tue Dec 18 06:32:11 2007
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to stephen.marquard@uct.ac.za using -f
From: stephen.marquard@uct.ac.za
Author: stephen.marquard@uct.ac.za
From stephen.marquard@uct.ac.za Tue Dec 18 06:31:24 2007
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to stephen.marquard@uct.ac.za using -f
From: stephen.marquard@uct.ac.za
Author: stephen.marquard@uct.ac.za
From stephen.marquard@uct.ac.za Sun Dec 16 03:59:13 2007
X-Authentication-Warning: nakamura.uits.iupui.edu: apache set sender to stephen.marquard@uct.ac.za using -f
From: stephen.marquard@uct.ac.za
Author: stephen.marquard@uct.ac.za
From stephen.marquard@uct.ac.za S

In [11]:
# letting the user choose the filename and get the count of number of subject lines
fname = input('Enter the file name')
fhand = open(fname)
count = 0
for line in fhand:
    line = line.strip()
    if line.startswith('Subject:'):
        count = count + 1
print(
'There were ',count,'subject lines in the text')

Enter the file namembox.txt
There were  1797 subject lines in the text


In [12]:
# using try,except and open
fname = input('Enter the file name')
try:
    fhand = open(fname)
except: 
    print('File cannot be opened:',fname)
    exit()
    
count = 0
for line in fhand:
    if line.startswith('Subject:'):
        count = count + 1
print('Number of lines starting with Subject: ',count)

Enter the file namembox.txt
Number of lines starting with Subject:  1797


In [13]:
# writing files
# if file already existing it clears the old data and start fresh
fout = open('output.txt','w')
print(fout)

<_io.TextIOWrapper name='output.txt' mode='w' encoding='cp1252'>


In [14]:
# write method does not add new line automatically
line1 = 'This here is the wattle,\n'
line2 = 'the emblem of our land,\n'
fout.write(line1)
fout.write(line2)

24

In [15]:
fout.close()

In [16]:
s = '1 2\t 3\n 4'
print(s)

1 2	 3
 4


In [17]:
# to view whitespace
print(repr(s))

'1 2\t 3\n 4'


### Exercises
#### Exercise 1
Write a program to read through a file  and print the contents of the file all in upper case.


In [18]:
fhand = open('mbox-short.txt')

In [19]:
for line in fhand:
    line = line.strip()
    line = line.upper()
    print(line)

FROM STEPHEN.MARQUARD@UCT.AC.ZA SAT JAN  5 09:14:16 2008
RETURN-PATH: <POSTMASTER@COLLAB.SAKAIPROJECT.ORG>
RECEIVED: FROM MURDER (MAIL.UMICH.EDU [141.211.14.90])
BY FRANKENSTEIN.MAIL.UMICH.EDU (CYRUS V2.3.8) WITH LMTPA;
SAT, 05 JAN 2008 09:14:16 -0500
X-SIEVE: CMU SIEVE 2.3
RECEIVED: FROM MURDER ([UNIX SOCKET])
BY MAIL.UMICH.EDU (CYRUS V2.2.12) WITH LMTPA;
SAT, 05 JAN 2008 09:14:16 -0500
RECEIVED: FROM HOLES.MR.ITD.UMICH.EDU (HOLES.MR.ITD.UMICH.EDU [141.211.14.79])
BY FLAWLESS.MAIL.UMICH.EDU () WITH ESMTP ID M05EEFR1013674;
SAT, 5 JAN 2008 09:14:15 -0500
RECEIVED: FROM PAPLOO.UHI.AC.UK (APP1.PROD.COLLAB.UHI.AC.UK [194.35.219.184])
BY HOLES.MR.ITD.UMICH.EDU ID 477F90B0.2DB2F.12494 ;
5 JAN 2008 09:14:10 -0500
RECEIVED: FROM PAPLOO.UHI.AC.UK (LOCALHOST [127.0.0.1])
BY PAPLOO.UHI.AC.UK (POSTFIX) WITH ESMTP ID 5F919BC2F2;
SAT,  5 JAN 2008 14:10:05 +0000 (GMT)
MESSAGE-ID: <200801051412.M05ECIAH010327@NAKAMURA.UITS.IUPUI.EDU>
MIME-VERSION: 1.0
CONTENT-TRANSFER-ENCODING: 7BIT
RECEIVED: FROM PR

- LEFT MOOT (UNUSED) ENTRIES COMMENTED FOR NOW

----------------------
THIS AUTOMATIC NOTIFICATION MESSAGE WAS SENT BY SAKAI COLLAB (HTTPS://COLLAB.SAKAIPROJECT.ORG/PORTAL) FROM THE SOURCE SITE.
YOU CAN MODIFY HOW YOU RECEIVE NOTIFICATIONS AT MY WORKSPACE > PREFERENCES.



FROM WAGNERMR@IUPUI.EDU FRI JAN  4 10:38:42 2008
RETURN-PATH: <POSTMASTER@COLLAB.SAKAIPROJECT.ORG>
RECEIVED: FROM MURDER (MAIL.UMICH.EDU [141.211.14.90])
BY FRANKENSTEIN.MAIL.UMICH.EDU (CYRUS V2.3.8) WITH LMTPA;
FRI, 04 JAN 2008 10:38:42 -0500
X-SIEVE: CMU SIEVE 2.3
RECEIVED: FROM MURDER ([UNIX SOCKET])
BY MAIL.UMICH.EDU (CYRUS V2.2.12) WITH LMTPA;
FRI, 04 JAN 2008 10:38:42 -0500
RECEIVED: FROM SHINING.MR.ITD.UMICH.EDU (SHINING.MR.ITD.UMICH.EDU [141.211.93.153])
BY FLAWLESS.MAIL.UMICH.EDU () WITH ESMTP ID M04FCFJM012313;
FRI, 4 JAN 2008 10:38:41 -0500
RECEIVED: FROM PAPLOO.UHI.AC.UK (APP1.PROD.COLLAB.UHI.AC.UK [194.35.219.184])
BY SHINING.MR.ITD.UMICH.EDU ID 477E52FA.E6C6E.24093 ;
4 JAN 2008 10:38:37 -0500
RECEIVED: 


#### Exercise 2
Write a program to prompt for a file name, and then read through the file and look for lines of the form:
X-DSPAM-Confidence: 0.8475

In [20]:
count = 0
total = 0
fname = input('Enter the filename')
fhand = open(fname)
for line in fhand:
    line = line.strip()
    if not line.startswith('X-DSPAM-Confidence'):
        continue
    line = line[-6:]
    line = float(line)
    count = count +1
    total = total + line
average = total/count
print(average)

Enter the filenamembox.txt
0.8941280467445736


#### Exercise 3
Modify the program that prompts the user for the file name so that it prints a funny message when the user types in the exact file name "na na boo boo". The program should behave normally for all other files. 

## Lists
### A list is a sequence


In [21]:
cheeses = ['cheddar','Edam','Gouda']
numbers = [17,123]
empty = []
print(cheeses, numbers, empty)

['cheddar', 'Edam', 'Gouda'] [17, 123] []


### Lists are mutable

In [22]:
print(cheeses[0])

cheddar


In [23]:
numbers[1] = 5
numbers

[17, 5]

In [24]:
'Edam' in cheeses

True

### Traversing a list

In [25]:
for cheese in cheeses:
    print(cheese)

cheddar
Edam
Gouda


In [26]:
# to wrtie to the list
for i in range(len(numbers)):
    numbers[i] = numbers[i]*2
    print(numbers[i])

34
10


In [27]:
for x in empty:
    print('this never happens')

In [28]:
neslst = ['spam',1, ['Brie','Roquefort','Pol leVeq'],[1,2,3]]
print(len(neslst))

4


### List operations

* '+' operator concatenates list
* '*' repeats the list a given number of times


In [29]:
a = [1,2,3]
b = [4,5,6]
c = a + b
print(c)

[1, 2, 3, 4, 5, 6]


In [30]:
[0] * 4

[0, 0, 0, 0]

### List slices


### List  methods

### Lists and functions

In [32]:
# compute the average of list of numbers entered by the user using a list

# without list
total = 0
count = 0
while(True):
    inp = input('Enter a number')
    if inp == 'done': break
    value = float(inp)
    total = total + value
    count = count + 1
average = total/count
print(average)

Enter a number3
Enter a number4
Enter a number23
Enter a numberdone
10.0


In [33]:
# using list
numlist = list()
while True:
    inp = input('Enter the number: ')
    if inp == 'done': break
    value = float(inp)
    numlist.append(value)
print(len(numlist))
average = sum(numlist) / len(numlist)
print('Average: ',average)

Enter the number: 5
Enter the number: 3
Enter the number: 45
Enter the number: 3
Enter the number: done
4
Average:  14.0


In [ ]:
n = [3,4,4,5,3,3]

In [ ]:
sum(n)

In [2]:
# Lists and strings

s = 'spam'
t = list(s)
t

['s', 'p', 'a', 'm']

In [3]:
s = 'pinnning for the fjords'
t = s.split()
t

['pinnning', 'for', 'the', 'fjords']

In [5]:
s = 'spam-spam-spam'
delimiter = '-'
s.split(delimiter)


['spam', 'spam', 'spam']

In [80]:
# join is the inverse of split
t = ['pining','for','the','fjords']
delimiter = ' '
delimiter.join(t)

'pining for the fjords'

In [35]:
# Parsing lines

fhand = open('mbox-short.txt')
for line in fhand:
    if not line.startswith('From'):
        continue
    wlist = line.split()
    if len(wlist) == 2:
        continue
    words = line.split()
    print(words[2])


Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu


In [36]:
# Objects and values

a = 'banana'
b = 'banana'
a is b

True

In [38]:
# Python created only one string object and both a & b refer to it

a = [1,2,3]
b = [1,2,3]
a is b

False

In [39]:
# Two lists are equivalent but are not identical

# Aliasing

a = [1,2,3]
b = a
a is b

True

In [40]:
# the association of a variable with an object is called reference

# object with more than one reference is said to be aliased

# changes made with one alias affect the other

b[0] = 17
a

[17, 2, 3]

In [41]:
# for string it never makes a difference whether a & b refer to the same string or not

List Arguments


In [46]:
# A reference of the list is passed to the function
def delete_head(t):
    del t[0]
    
letter = ['a','b','c','d']
letter


['a', 'b', 'c', 'd']

In [47]:
delete_head(letter)

In [48]:
letter

['b', 'c', 'd']

In [79]:
# append method modifies the list, + operator creates a new list

t1 = [1,2]
t2 = t1.append(3)
t1

[1, 2, 3]

In [54]:
print(t2)

None


In [56]:
t3 = t1 + [3]
t3

[1, 2, 3, 3]

In [57]:
t3

[1, 2, 3, 3]

In [59]:
t2 is t3

False

In [65]:
# below function does not delete the head of a list

def bad_delete_head(t):
    t = t[1:]  # wrong
    
def tail(t):
    return t[1:]

letters = ['a','b','c']
rest = tail(letters)
print(rest)


['b', 'c']


### Exercis 1
Write a function called chop that takes a list and modifies it, removing the first and last elements, and returns NOne. Then write a function called middle that takes a list and returns a new list that contains all but the first and last elements


In [76]:
# writing a function called chop

def chop(t):
    t = t[1:]
    t = t[:-2]
def middle(t):
    t = t[1:-1]
    return t
letters = ['a','b','c','d']
print(chop(letters))

None


In [77]:
print(middle(letters))

['b', 'c']


## Debugging

#### Most list methods modify the argument  and return None. This is the opposite of string methods which return a new string leave the orgiginal alone

In [82]:
t = t.sort() # WRONG! returns None

In [85]:
t = ['fjords','for','pining','the']

In [86]:
t.append(x)

['fjords', 'for', 'pining', 'the']

In [88]:
x = 5
t.append(x)
t = t + [x]

#Wrong:
t.append([x])
t = t.append(x)
t + x
t = t + x

TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'

In [2]:
t = ['fjords','for','pining','the']
tsrt = sorted(t)
tsrt

['fjords', 'for', 'pining', 'the']

In [14]:
# parsing lines using list

fhand = open('mbox.txt')
for line in fhand:

    words = line.split()
    print('debug:',words)
    if words[0] != 'From':
        continue
    print(words)
    print(words[2])

debug: ['From', 'stephen.marquard@uct.ac.za', 'Sat', 'Jan', '5', '09:14:16', '2008']
['From', 'stephen.marquard@uct.ac.za', 'Sat', 'Jan', '5', '09:14:16', '2008']
Sat
debug: ['Return-Path:', '<postmaster@collab.sakaiproject.org>']
debug: ['Received:', 'from', 'murder', '(mail.umich.edu', '[141.211.14.90])']
debug: ['by', 'frankenstein.mail.umich.edu', '(Cyrus', 'v2.3.8)', 'with', 'LMTPA;']
debug: ['Sat,', '05', 'Jan', '2008', '09:14:16', '-0500']
debug: ['X-Sieve:', 'CMU', 'Sieve', '2.3']
debug: ['Received:', 'from', 'murder', '([unix', 'socket])']
debug: ['by', 'mail.umich.edu', '(Cyrus', 'v2.2.12)', 'with', 'LMTPA;']
debug: ['Sat,', '05', 'Jan', '2008', '09:14:16', '-0500']
debug: ['Received:', 'from', 'holes.mr.itd.umich.edu', '(holes.mr.itd.umich.edu', '[141.211.14.79])']
debug: ['by', 'flawless.mail.umich.edu', '()', 'with', 'ESMTP', 'id', 'm05EEFR1013674;']
debug: ['Sat,', '5', 'Jan', '2008', '09:14:15', '-0500']
debug: ['Received:', 'FROM', 'paploo.uhi.ac.uk', '(app1.prod.collab

IndexError: list index out of range

In [18]:
fhand = open('mbox.txt')
for line in fhand:

    words = line.split()
    #print('debug:',words)
    if len(words) == 0:
        continue
    if words[0] != 'From':
        continue
    print(words[2])

Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Wed
Tue
Tue
Mon
Mon
Mon
Mon
Mon
Mon
Mon
Sun
Sun
Sun
Sun
Sun
Sun
Sun
Sat
Sat
Sat
Sat
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Wed
Wed
Wed
Tue
Mon
Mon
Sun
Sat
Sat
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Fri
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Thu
Wed
Wed
Wed
Wed
Wed
Wed
Wed


### Exercise 2

Figure out which line of the above program  is still not properly guarded. see if you can cosntruct a text file which causes the program to fail and then modify the program so that the line is properly guarded and test it to make sure it handles your new text file.



### Exercise 3
Rewrite the guardian code in the above example without two if statements. Instead,use a compound logical expression using the and logical operator with a single if statement

#### Exercise 4

Download a copy of Romeo.txt. Write a program to open the file romeo.txt and read it line by line. For each line, split the line into a list of words using the split function. For each word check to see if the word is already in a list. if the word is not in the list add it to the list. When the program completes, sort and print the resulting words in alphabetical order.


### Exercise 5

Write a program to read through the mail box data and when you find line that starts with "From". you will split the line into words using the split function. We are interested in who sent the message, which is the second word on the From line.You will parse the from line and print out the second word for each from line. then you will also count the number of from lines and print out a count at the end. 

### Exercise 6
Rewrite the program that prompts the user for a list of numbers and prints out the maximum and minimum of the numbers at the end when the user
enters done. Write the program to store the numbers the user enters in a list and use the max() and min() function to compute the maximum 
and minimum numbers after the loop completes.

## Dictionaries

In [22]:
eng2sp = dict()
eng2sp

{}

In [23]:
eng2sp['one'] = 'uno'
eng2sp

{'one': 'uno'}

In [24]:
eng2sp = {'one':'uno','two':'dos','three':'tres'}
eng2sp

{'one': 'uno', 'two': 'dos', 'three': 'tres'}

In [25]:
# order of items is unpredictable
print(eng2sp['two'])

dos


In [26]:
print(eng2sp['fsd'])

KeyError: 'fsd'

In [27]:
# return number of key value pairs

print(len(eng2sp))

3


In [29]:
# in operator checks existance of key in the item
'one' in eng2sp

True

In [30]:
'ten' in eng2sp

False

In [31]:
# to check value in item

vals = list(eng2sp.values())
vals

['uno', 'dos', 'tres']

In [32]:
'uno' in vals

True

* in operator may take more time for searching list and same time for dictionary(hash table)

### Exercise 1:
Write a program that reads words in words.txt and stores them as keys in a dictionary. It doesnt matter what the values are. Then you can use the in operator to check whether a string is in the dictionary.

### Dictionary as a set of counters

You could create a dictionary with characters as keys and counters as the corresponding values. The first time you see a character, you would add an item to the dictionary. After that you would increment the value of an existing item. 


In [33]:
word = 'brontasaurus'
d = dict()
for c in word:
    if c not in d:
        d[c]=1
    else:
        d[c] = d[c]+1
print(d)

{'b': 1, 'r': 2, 'o': 1, 'n': 1, 't': 1, 'a': 2, 's': 2, 'u': 2}


In [35]:
# get method takes a key and a default value

counts = {'chunk':1,'annie':42,'jan':100}
counts.get('annie')

42

In [36]:
counts.get('annie3',0)

0

In [37]:
# counter using get

word = 'brontosaurus'
d = dict()
for c in word:
    d[c]=d.get(c,0) + 1
print(d)


{'b': 1, 'r': 2, 'o': 2, 'n': 1, 't': 1, 's': 2, 'a': 1, 'u': 2}


### Dictionaries and files

In [1]:
fname = input('enter the filename: ')
try:
    fhand = open(fname)
except:
    print('file cannot be opened')
    exit()
    
counts = dict()
for line in fhand:
    words = line.split()
    for word in words:
        if word not in counts:
            counts[word] = 1
        else:
            counts[word] += 1
print(counts)

enter the filename: romeo.txt
{'But': 1, 'soft': 1, 'what': 1, 'light': 1, 'through': 1, 'yonder': 1, 'window': 1, 'breaks': 1, 'It': 1, 'is': 3, 'the': 3, 'east': 1, 'and': 3, 'Juliet': 1, 'sun': 2, 'Arise': 1, 'fair': 1, 'kill': 1, 'envious': 1, 'moon': 1, 'Who': 1, 'already': 1, 'sick': 1, 'pale': 1, 'with': 1, 'grief': 1}


### Looping and dictionaries

In [2]:
counts = {'chuck':1,'annie':42,'jan':100}
counts

{'chuck': 1, 'annie': 42, 'jan': 100}

In [3]:
for key in counts:
    print(key, counts[key])

chuck 1
annie 42
jan 100


In [5]:
for key in counts:
    if counts[key]>10:
        print(key, counts[key])
    

annie 42
jan 100


In [7]:
# print dictionary in alphabetical order

counts = {'chuck':1, 'annie':42, 'jan':100}
lst = list(counts.keys())
lst.sort()
for key in lst:
    print(key,counts[key])


annie 42
chuck 1
jan 100


### Advanced text parsing


In [8]:
# lower, punctuation, translate
import string
string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [8]:
import string
fname = input('Enter the file name: ')
try:
    fhand = open(fname)
except:
    print('File cannot be opened:',fname)
    exit()

counts = dict()

for line in fhand:
    #print(line)
    line = line.rstrip()
    line = line.translate(line.maketrans('','',string.punctuation))
   
    line = line.lower()
    #line = line.strip()
    words = line.split()
    for word in words:
        if word not in counts:
            counts[word] = 1
        else:
            counts[word] += 1 
    #print(line)
print(counts)


Enter the file name: romeo-full.txt
{'romeo': 40, 'and': 42, 'juliet': 32, 'act': 1, '2': 2, 'scene': 2, 'ii': 1, 'capulets': 1, 'orchard': 2, 'enter': 1, 'he': 5, 'jests': 1, 'at': 9, 'scars': 1, 'that': 30, 'never': 2, 'felt': 1, 'a': 24, 'wound': 1, 'appears': 1, 'above': 6, 'window': 2, 'but': 18, 'soft': 1, 'what': 11, 'light': 5, 'through': 2, 'yonder': 2, 'breaks': 1, 'it': 22, 'is': 21, 'the': 34, 'east': 1, 'sun': 2, 'arise': 1, 'fair': 4, 'kill': 2, 'envious': 2, 'moon': 4, 'who': 5, 'already': 1, 'sick': 2, 'pale': 1, 'with': 8, 'grief': 2, 'thou': 32, 'her': 14, 'maid': 2, 'art': 7, 'far': 2, 'more': 9, 'than': 6, 'she': 9, 'be': 14, 'not': 18, 'since': 1, 'vestal': 1, 'livery': 1, 'green': 1, 'none': 1, 'fools': 1, 'do': 7, 'wear': 1, 'cast': 1, 'off': 1, 'my': 29, 'lady': 2, 'o': 11, 'love': 24, 'knew': 1, 'were': 9, 'speaks': 3, 'yet': 9, 'says': 1, 'nothing': 1, 'of': 20, 'eye': 2, 'discourses': 1, 'i': 61, 'will': 8, 'answer': 1, 'am': 7, 'too': 8, 'bold': 1, 'tis': 4,

### Exercise 2:
Write a program that categorizes each mail message by
which day of the week the commit was done. To do this look for lines
that start with “From”, then look for the third word and keep a running
count of each of the days of the week. At the end of the program print
out the contents of your dictionary (order does not matter).

### Exercise 3: 
Write a program to read through a mail log, build a his-
togram using a dictionary to count how many messages have come from
each email address, and print the dictionary.

### Exercise 4: 
Add code to the above program to figure out who has the
most messages in the file. After all the data has been read and the dic-
tionary has been created, look through the dictionary using a maximum
loop (see Chapter 5: Maximum and minimum loops) to find who has
the most messages and print how many messages the person has.

### Exercise 5: 
This program records the domain name (instead of the
address) where the message was sent from instead of who the mail came
from (i.e., the whole email address). At the end of the program, print
out the contents of your dictionary.

## Tuples

In [9]:
#Tuples are immutable

t = 'a','b','c','d','e'
t

('a', 'b', 'c', 'd', 'e')

In [11]:
t = ('a','b','c','d','e')
type(t)

tuple

In [14]:
# to create tuple with single element we have to include final comma
t1 = ('a',)
type(t1)

tuple

In [15]:
# another way

t = tuple()
print(t)

()


In [20]:
t = tuple('lupins')
t

('l', 'u', 'p', 'i', 'n', 's')

In [21]:
t = ('a','b','c','d','e')
print(t[0])

a


In [22]:
print(t[1:3])

('b', 'c')


In [23]:
# cannot modify element
t[0]='dummy'

TypeError: 'tuple' object does not support item assignment

In [24]:
# replace one tuple with another
t = ('A',) + t[1:]
t

('A', 'b', 'c', 'd', 'e')

### comparing tuples

In [25]:
(0,1,2) < (0,3,4)

True

In [26]:
(0,1,2000) < (0,3,4)

True

In [37]:
# sort list of words from longest to shortest

txt = 'but soft what light in yander window breaks'
words = txt.split()
t = list()
for word in words:
    t.append((len(word),word))
t.sort(reverse = True)

res = list()
for length,word in t:
    res.append(word)
print(res)

['yander', 'window', 'breaks', 'light', 'what', 'soft', 'but', 'in']


### Tuple assignment

In [1]:
m = ['have','fun']
x, y = m
x

'have'

In [2]:
y

'fun'

In [3]:
(x,y) = m
x,y

('have', 'fun')

In [5]:
# swap values of a variable
x,y = y,x
x,y

('have', 'fun')

In [6]:
# split email address into username and domain
addr = 'monty@python.org'

uname, domain = addr.split('@')
uname, domain

('monty', 'python.org')

In [7]:
addr = 'monty@python.org'

addrlst = addr.split('@')
addrlst

['monty', 'python.org']

In [8]:
a, b = ['niju', 'shanth']
a,b

('niju', 'shanth')

### Dictionary and Tuples


In [15]:
d = {'b':20,'a':1,'c':33}
d

{'b': 20, 'a': 1, 'c': 33}

In [16]:
t = list(d.items())
print(t)

[('b', 20), ('a', 1), ('c', 33)]


In [17]:
t.sort()
t

[('a', 1), ('b', 20), ('c', 33)]

### Multiple assignemnt with dictionaries

In [25]:
for key, val in list(d.items()):
    print(key,val)


b 20
a 1
c 33


In [33]:
# print in sorted order
l = list()
for key, val in list(d.items()):
    l.append((val,key))
l

[(20, 'b'), (1, 'a'), (33, 'c')]

In [36]:
l.sort(reverse = True)
l

[(33, 'c'), (20, 'b'), (1, 'a')]

### The most common words



In [47]:
# print 10 most common words in the text
import string
fhand = open('romeo-full.txt')
counts = dict()
for line in fhand:
    line = line.translate(str.maketrans('','',string.punctuation))
    line = line.lower()
    words = line.split()
    for word in words:
        if word not in counts:
            counts[word] = 1
        else:
            counts[word] += 1
#print(counts)

# sort the dictionary by value

lst = list()
for key,val in list(counts.items()):
    lst.append((val,key))
lst.sort(reverse = True)
for key,val in lst[:10]:
    print(key,val)

61 i
42 and
40 romeo
34 to
34 the
32 thou
32 juliet
30 that
29 my
24 thee


### Using tuples as keys in dictionary



In [55]:
number = 6
last = 'niju'
first = 'geo'
directory = {(last,first):30}
print(directory)
directory[last,first] = number
for last,first in directory:
    print(first, last, directory[last,first])

{('niju', 'geo'): 30}
geo niju 6


### Sequences: strings, lists, and tuples - oh My!
#### Debugging


#### Exercise 1: 
Revise a previous program as follows: Read and parse the
“From” lines and pull out the addresses from the line. Count the num-
ber of messages from each person using a dictionary.
After all the data has been read, print the person with the most commits
by creating a list of (count, email) tuples from the dictionary. Then
sort the list in reverse order and print out the person who has the most
commits.

#### Exercise 2: 
This program counts the distribution of the hour of the day
for each of the messages. You can pull the hour from the “From” line
by finding the time string and then splitting that string into parts using
the colon character. Once you have accumulated the counts for each
hour, print out the counts, one per line, sorted by hour as shown below

#### Exercise 3: 
Write a program that reads a file and prints the letters
in decreasing order of frequency. Your program should convert all the
input to lower case and only count the letters a-z. Your program should
not count spaces, digits, punctuation, or anything other than the letters
a-z. Find text samples from several different languages and see how
letter frequency varies between languages. Compare your results with
the tables at https://wikipedia.org/wiki/Letter_frequencies.

## Chapter 11 Regular Expressions


In [57]:
# search function

import re
    
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    if re.search('From:',line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [58]:
# search for lines that start with from

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.strip()
    if re.search('^From:',line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


## character matching in regular expressions


In [60]:
# search for lines that starts with 'F', followed by 2 characters followed by 'm'

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.strip()
    if re.search('^F..m:',line):
        print(line)

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


In [66]:
# search for lines that start with from and have an at sign

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    if re.search('^From:.+@',line):
        print(line)
    

From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: zqian@umich.edu
From: rjlowe@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: gsilver@umich.edu
From: gsilver@umich.edu
From: zqian@umich.edu
From: gsilver@umich.edu
From: wagnermr@iupui.edu
From: zqian@umich.edu
From: antranig@caret.cam.ac.uk
From: gopal.ramasammycook@gmail.com
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: david.horwitz@uct.ac.za
From: stephen.marquard@uct.ac.za
From: louis@media.berkeley.edu
From: louis@media.berkeley.edu
From: ray@media.berkeley.edu
From: cwen@iupui.edu
From: cwen@iupui.edu
From: cwen@iupui.edu


### Extracting data using regular expressions

In [69]:
# find email address from line

import re
s = 'A message from csev@umich.edu to cwen@iupi.edu about meeting @2pm'
lst = re.findall('\S+@\S+',s) #returns a list of found objects
print(lst)

['csev@umich.edu', 'cwen@iupi.edu']


In [74]:
# search for lines that have an at sign between characters
import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('\S+@\S+',line)
    if len(x)>0:
        print(x)
    

['stephen.marquard@uct.ac.za']
['<postmaster@collab.sakaiproject.org>']
['<200801051412.m05ECIaH010327@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['apache@localhost)']
['source@collab.sakaiproject.org;']
['stephen.marquard@uct.ac.za']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['stephen.marquard@uct.ac.za']
['louis@media.berkeley.edu']
['<postmaster@collab.sakaiproject.org>']
['<200801042308.m04N8v6O008125@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['<source@collab.sakaiproject.org>;']
['apache@localhost)']
['source@collab.sakaiproject.org;']
['louis@media.berkeley.edu']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['louis@media.berkeley.edu']
['zqian@umich.edu']
['<postmaster@collab.sakaiproject.org>']
['<200801042109.m04L92hb007923@nakamura.uits.iupui.edu>']
['<source@collab.sakaiproject

In [77]:
# search for lines that have an at sign between characters
# the characters must be a letter or number

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('[a-zA-Z0-9]\S+@\S+[a-xA-Z]',line)
    if len(x)>0:
        print(x)

['stephen.marquard@uct.ac.za']
['postmaster@collab.sakaiproject.org']
['200801051412.m05ECIaH010327@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['apache@localhost']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['source@collab.sakaiproject.org']
['stephen.marquard@uct.ac.za']
['stephen.marquard@uct.ac.za']
['louis@media.berkeley.edu']
['postmaster@collab.sakaiproject.org']
['200801042308.m04N8v6O008125@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject.org']
['apache@localhost']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['source@collab.sakaiproject.org']
['louis@media.berkeley.edu']
['louis@media.berkeley.edu']
['zqian@umich.edu']
['postmaster@collab.sakaiproject.org']
['200801042109.m04L92hb007923@nakamura.uits.iupui.edu']
['source@collab.sakaiproject.org']
['source@collab.sakaiproject

### Combining, searching, extracting

In [89]:
# find numbers on line that starts with X

# search for lines that starts with'X' followed by any non white space character and ':'
# followed by space and any number
# the number can include decimal

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
#     if re.search('^X\S*: [0-9.]+',line):
#         print(line)
    x = re.findall('^X\S*: ([0-9.]+)',line)
    if len(x)>0:
        print(x)

['0.8475']
['0.0000']
['0.6178']
['0.0000']
['0.6961']
['0.0000']
['0.7565']
['0.0000']
['0.7626']
['0.0000']
['0.7556']
['0.0000']
['0.7002']
['0.0000']
['0.7615']
['0.0000']
['0.7601']
['0.0000']
['0.7605']
['0.0000']
['0.6959']
['0.0000']
['0.7606']
['0.0000']
['0.7559']
['0.0000']
['0.7605']
['0.0000']
['0.6932']
['0.0000']
['0.7558']
['0.0000']
['0.6526']
['0.0000']
['0.6948']
['0.0000']
['0.6528']
['0.0000']
['0.7002']
['0.0000']
['0.7554']
['0.0000']
['0.6956']
['0.0000']
['0.6959']
['0.0000']
['0.7556']
['0.0000']
['0.9846']
['0.0000']
['0.8509']
['0.0000']
['0.9907']
['0.0000']


In [90]:
# search for lines that start with 'Details: rev='
# followed by numbers and '.'
# Then print the numbe if it is greater than zero

import re
hand = open('mbox-short.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('^Details:.*rev=([0-9.]+)',line)
    if len(x)>0:
        print(x)

['39772']
['39771']
['39770']
['39769']
['39766']
['39765']
['39764']
['39763']
['39762']
['39761']
['39760']
['39759']
['39758']
['39757']
['39756']
['39755']
['39754']
['39753']
['39752']
['39751']
['39750']
['39749']
['39746']
['39745']
['39744']
['39743']
['39742']


In [91]:
# search for ines that start with from and a character
# followed by a two digit number between 00 and 99 followed by ':'
# then print the number if it is greater than zero

import re
hand = open('mbox.txt')
for line in hand:
    line = line.rstrip()
    x = re.findall('^From .* ([0-9][0-9]):',line)
    if len(x)>0:
        print(x)

['09']
['18']
['16']
['15']
['15']
['14']
['11']
['11']
['11']
['11']
['11']
['11']
['10']
['10']
['10']
['09']
['07']
['06']
['04']
['04']
['04']
['19']
['17']
['17']
['16']
['16']
['16']
['15']
['15']
['15']
['15']
['15']
['14']
['13']
['13']
['13']
['13']
['13']
['13']
['13']
['13']
['13']
['12']
['12']
['12']
['12']
['11']
['11']
['11']
['10']
['10']
['10']
['10']
['10']
['10']
['10']
['10']
['09']
['09']
['09']
['09']
['09']
['08']
['18']
['18']
['17']
['17']
['17']
['16']
['16']
['15']
['15']
['15']
['14']
['12']
['12']
['11']
['11']
['09']
['09']
['09']
['09']
['09']
['09']
['08']
['08']
['08']
['08']
['07']
['05']
['04']
['04']
['03']
['03']
['03']
['03']
['02']
['20']
['07']
['16']
['16']
['14']
['14']
['14']
['10']
['09']
['22']
['16']
['16']
['13']
['12']
['10']
['10']
['21']
['14']
['10']
['10']
['23']
['16']
['14']
['12']
['12']
['11']
['21']
['21']
['17']
['17']
['17']
['17']
['17']
['13']
['12']
['10']
['10']
['10']
['09']
['09']
['09']
['09']
['08']
['21']
['16']
['11']

['06']
['06']
['00']
['23']
['23']
['23']
['14']
['14']
['13']
['12']
['12']
['11']
['11']
['11']
['10']
['10']
['10']
['10']
['09']
['09']
['09']
['09']
['08']
['08']
['08']
['08']
['08']
['07']
['07']
['07']
['06']
['06']
['06']
['16']
['15']
['15']
['15']
['15']
['15']
['11']
['10']
['10']
['10']
['10']
['09']
['09']
['09']
['09']
['09']
['09']
['09']
['08']
['08']
['08']
['08']
['08']
['07']
['07']
['07']
['07']
['07']
['07']
['07']
['06']
['06']
['06']
['17']
['17']
['17']
['17']
['17']
['17']
['17']
['17']
['17']
['17']
['16']
['16']
['16']
['16']
['16']
['16']
['16']
['16']
['16']
['16']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['15']
['14']
['14']
['14']
['14']
['13']
['13']
['13']
['13']
['13']
['13']
['13']
['12']
['12']
['12']
['12']
['12']
['12']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['11']
['10']
['10']
['10']
['10']
['10']
['09']
['09']
['09']
['04']
['22']
['22']
['20']
['19']
['18']
['17']
['17']

### Escape character



In [2]:
import re
x = 'We just received $10.00 for cookies.'
y = re.findall('\$[0-9.]+',x) # inside square bracket, characters are not special
#y = re.findall('[$0-9.]+',x) # inside square bracket, characters are not special
print(y)

['$10.00']


* ^ matches the beginning of the line
* $ matches the end of the line
* . matches any character
* \s matches a whitespace character
* \S matches a non whitespace character
* \* applies to the immediately preceding character and indicates to match zero or more
* *? applies to the immediately preceding character and indicates to match zero or more in non greedy mode
* \+ Applies to the immediately preceding character(s) and indicates to match one or
more times.
* +? Applies to the immediately preceding character(s) and indicates to match one
    or more times in “non-greedy mode”.
* ? Applies to the immediately preceding character(s) and indicates to match zero
or one time.
* ?? Applies to the immediately preceding character(s) and indicates to match zero
or one time in “non-greedy mode”.
* [aeiou] Matches a single character as long as that character is in the specified set.
In this example, it would match “a”, “e”, “i”, “o”, or “u”, but no other characters.
* [a-z0-9] You can specify ranges of characters using the minus sign. This example
is a single character that must be a lowercase letter or a digit.
* [ˆA-Za-z] When the first character in the set notation is a caret, it inverts the
logic. This example matches a single character that is anything other than an
uppercase or lowercase letter.
* ( ) When parentheses are added to a regular expression, they are ignored for the
purpose of matching, but allow you to extract a particular subset of the matched
string rather than the whole string when using findall().
* \b Matches the empty string, but only at the start or end of a word.
* \B Matches the empty string, but not at the start or end of a word.
* \d Matches any decimal digit; equivalent to the set [0-9].
* \D Matches any non-digit character; equivalent to the set [ˆ0-9].

In [3]:
help(re.search)

Help on function search in module re:

search(pattern, string, flags=0)
    Scan through string looking for a match to the pattern, returning
    a Match object, or None if no match was found.



### Exercise 1
Write a simple program to simulate the operation of the
grep command on Unix. Ask the user to enter a regular expression and
count the number of lines that matched the regular expression

### Exercise 2: 
Write a program to look for lines of the form:
New Revision: 39772
Extract the number from each of the lines using a regular expression
and the findall() method. Compute the average of the numbers and
print out the average as an integer.

## Chapter 12 
## Networked programs

### The worlds simpleast web browser


In [8]:
# request a document and display what the server sends back

import socket
mysock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
mysock.connect(('data.pr4e.org',80))
cmd = 'GET http://data.pr4e.org/romeo.txt HTTP/1.0\r\n\r\n'.encode()
mysock.send(cmd)

while True:
    data = mysock.recv(512)
    if len(data) < 1:
        break
    print(data.decode(),end = '')
mysock.close()

HTTP/1.1 200 OK
Date: Sat, 22 Feb 2020 07:18:28 GMT
Server: Apache/2.4.18 (Ubuntu)
Last-Modified: Sat, 13 May 2017 11:22:22 GMT
ETag: "a7-54f6609245537"
Accept-Ranges: bytes
Content-Length: 167
Cache-Control: max-age=0, no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: Wed, 11 Jan 1984 05:00:00 GMT
Connection: close
Content-Type: text/plain

But soft what light through yonder window breaks
It is the east and Juliet is the sun
Arise fair sun and kill the envious moon
Who is already sick and pale with grief


In [9]:
b'Hello world'

b'Hello world'

In [10]:
'Hello world'.encode()

b'Hello world'

### Retrieving and image over HTTP


In [14]:
import socket
import time

HOST = 'data.pr4e.org'
PORT = 80
mysock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)
mysock.connect((HOST,PORT))
mysock.sendall(b'GET http://data.pr4e.org/cover3.jpg HTTP/1.0\r\n\r\n')
count = 0
picture = b""

while True:
    data = mysock.recv(5120)
    if len(data) < 1: break
    time.sleep(0.25)
    count = count + len(data)
    print(len(data),count)
    picture = picture + data
    
mysock.close()

# Look for the end of the header (2 CRLF)

pos = picture.find(b"\r\n\r\n")
print('Header length', pos)
print(picture[:pos].decode())

#Skip past the header and save the picture data

picture = picture[pos+4:]
fhand = open("stuff.jpg", 'wb')
fhand.write(picture)
fhand.close()

                   

5120 5120
5120 10240
5120 15360
5120 20480
5120 25600
5120 30720
5120 35840
5120 40960
5120 46080
5120 51200
5120 56320
5120 61440
5120 66560
5120 71680
5120 76800
5120 81920
5120 87040
5120 92160
5120 97280
5120 102400
5120 107520
5120 112640
5120 117760
5120 122880
5120 128000
5120 133120
5120 138240
5120 143360
5120 148480
5120 153600
5120 158720
5120 163840
5120 168960
5120 174080
5120 179200
5120 184320
5120 189440
5120 194560
5120 199680
5120 204800
5120 209920
5120 215040
5120 220160
5120 225280
5120 230400
208 230608
Header length 394
HTTP/1.1 200 OK
Date: Sat, 22 Feb 2020 07:38:14 GMT
Server: Apache/2.4.18 (Ubuntu)
Last-Modified: Mon, 15 May 2017 12:27:40 GMT
ETag: "38342-54f8f2e5b6277"
Accept-Ranges: bytes
Content-Length: 230210
Vary: Accept-Encoding
Cache-Control: max-age=0, no-cache, no-store, must-revalidate
Pragma: no-cache
Expires: Wed, 11 Jan 1984 05:00:00 GMT
Connection: close
Content-Type: image/jpeg


### Retrieving web pages with urllib

In [18]:
import urllib.request

fhand  = urllib.request.urlopen('http://data.pr4e.org/romeo.txt')
for line in fhand:
    print(line.strip())
    #print(line.decode().strip())

b'But soft what light through yonder window breaks'
b'It is the east and Juliet is the sun'
b'Arise fair sun and kill the envious moon'
b'Who is already sick and pale with grief'


In [24]:
# read from website and compute word ffrequency

import urllib.request, urllib.parse, urllib.error

fhand = urllib.request.urlopen('http://data.pr4e.org/romeo.txt')

counts = dict()
for line in fhand:
    words = line.decode().split()
    for word in words:
        counts[word] = counts.get(word,0) + 1

print(counts)
     


{'But': 1, 'soft': 1, 'what': 1, 'light': 1, 'through': 1, 'yonder': 1, 'window': 1, 'breaks': 1, 'It': 1, 'is': 3, 'the': 3, 'east': 1, 'and': 3, 'Juliet': 1, 'sun': 2, 'Arise': 1, 'fair': 1, 'kill': 1, 'envious': 1, 'moon': 1, 'Who': 1, 'already': 1, 'sick': 1, 'pale': 1, 'with': 1, 'grief': 1}


### Reading binary files using urllib


In [28]:
import urllib.request, urllib.parse, urllib.error

img = urllib.request.urlopen('http://data.pr4e.org/cover3.jpg').read()
fhand = open('cover3.jpg','wb')
fhand.write(img)
fhand.close()

In [35]:
# reading and writing in blocks to avoid out of memory

import urllib.request, urllib.parse, urllib.error

img = urllib.request.urlopen('http://data.pr4e.org/cover3.jpg')
fhand = open('cover3.jpg','wb')
size = 0
while True:
    info = img.read(100000)
    if len(info) < 1: break
    size = size + len(info)
    fhand.write(info)

print(size,'characters copied')
fhand.close()

230210 characters copied


### Parsing the html and scraping the web

### Parsing HTML using regular expression

In [37]:
# search the link values within URL input 

import urllib.request, urllib.parse, urllib.error
import re
import ssl

# Ignore SSL certificate errors

ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

url = input('Enter - ')
html = urllib.request.urlopen(url).read()
links = re.findall(b'href="(http[s]?://.*?)"',html)

for link in links:
    print(link.decode())

Enter - https://www.w3.org/Protocols/rfc2616/rfc2616.txt
